File: composite/java/Folder.java
package androidhot.github.io.designpatterndemo.composite.java;

import java.util.ArrayList;

public class Folder extends AbstractFile {
    private ArrayList<AbstractFile> abstractFiles = new ArrayList<>();
    private String folderName;

    public Folder(String folderName) {
        this.folderName = folderName;
    }

    @Override public void add(AbstractFile abstractFile) {
        abstractFiles.add(abstractFile);
    }


    @Override public void remove(AbstractFile abstractFile) {
        abstractFiles.remove(abstractFile);
    }


    @Override public AbstractFile getChild(int index) {
        return abstractFiles.get(index);
    }


    @Override public void killVirus() {
        System.out.println("Kill folder " + folderName + " virus.");
        for (AbstractFile file: abstractFiles) {
            file.killVirus();
        }
    }
}

File: composite/kotlin/Folder.kt
package androidhot.github.io.designpatterndemo.composite.kotlin

class Folder(private val folderName: String) : AbstractFile() {
    private var abstractFiles = ArrayList<AbstractFile>()
    override fun add(abstractFile: AbstractFile) {
        abstractFiles.add(abstractFile)
    }

    override fun remove(abstractFile: AbstractFile) {
        abstractFiles.remove(abstractFile)
    }

    override fun getChild(index: Int): AbstractFile {
        return abstractFiles[index]
    }

    override fun killVirus() {
        println("Kill folder $folderName virus.")
        for (abstractFile in abstractFiles) {
            abstractFile.killVirus()
        }
    }
}
